BBPebbleLib/Pebble.cpp

```diff
@@ -38,46 +44,88 @@ void Pebble::onDisconnect()
 
 void Pebble::onDataReceived(QBluetoothSocket &bt_socket)
 {
-    QDataStream in(&bt_socket);
-
-    if (this->payloadSize == 0) {
-        if (bt_socket.bytesAvailable() < (int)2*sizeof(quint16)) {
+    static const qint64 s_header_length = (int)2*sizeof(quint16);
+
+    if (bt_socket.bytesAvailable() >= s_header_length) {
+        // Take a look at the header, but do not remove it from the socket input buffer.
+        // We will only remove it once we're sure the entire packet is in the buffer.
+        uchar header[s_header_length];
+        bt_socket.peek(reinterpret_cast<char*>(header), s_header_length);
+
+        const quint16 message_length = qFromBigEndian<quint16>(&header[0]);
+        const quint16 endpoint = qFromBigEndian<quint16>(&header[2]);
+
+        // Sanity checks on the message_length
+        if (message_length == 0) {
+            qDebug() << "received empty message";
+            bt_socket.read(s_header_length); // skip this header
+            QTimer::singleShot(0, this, SLOT(onDataReceived(bt_socket))); // check if there are additional headers.
+            return;
+        } else if (message_length > 8 * 1024) {
+            // Protocol does not allow messages more than 8K long, seemingly.
+            qDebug() << "received message size too long: " << message_length;
+            bt_socket.read(message_length); // skip the too big message and look for another header
+            QTimer::singleShot(0, this, SLOT(onDataReceived(bt_socket))); // check if there are additional headers.
```

Comment:
i understood you wanted to return to the eventloop as soon as possible so i appended the slot via QTimer. Do you prefer QInvokeMethod? Or direct call? Im always open for feedback and improvements =)